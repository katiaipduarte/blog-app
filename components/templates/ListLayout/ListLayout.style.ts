import styled from 'styled-components';

export const ListContainer = styled.section`
  .title {
    padding: 1.5rem 0;
    border-bottom: 2px solid ${({ theme }) => theme.text};

    h1 {
      line-height: 2.25rem;
      font-weight: 800;
      font-size: 1.8rem;
    }

    p {
      font-size: 1.25rem;
      line-height: 1.75rem;
      color: ${({ theme }) => theme.grey};
    }
  }

  ul {
    li {
      padding: 1.5rem 0 3rem;

      &:not(&:last-child):not(&.blog-page) {
        border-bottom: 2px solid ${({ theme }) => theme.text};
      }

      article {
        dl {
          dd {
            line-height: 1.5rem;
            font-size: 1rem;
            color: ${({ theme }) => theme.grey};
          }
        }
        .post {
          .description {
            h2 {
              font-wight: 700;
              line-height: 2rem;
              font-size: 1.5rem;
              letter-spacing: 0.025rem;
              color: ${({ theme }) => theme.primaryColor};
            }

            p {
              color: ${({ theme }) => theme.grey};
              font-size: 1rem;
              line-height: 1.75;
              margin-top: 1.5rem;
            }
          }

          button {
            margin-top: 1rem;
          }
        }
      }
    }
  }

  .all-post-btn {
    float: right;
  }

  @media all and (min-width: 700px) {
    .title {
      padding: 2rem 0;

      h1 {
        line-height: 3.5rem;
        font-size: 3.75rem;
      }

      p {
        margin-top: 1.5rem;
      }
    }

    article {
      display: grid;
      grid-template-columns: repeat(4, minmax(0, 1fr));
      align-items: baseline;
    }

    .post {
      grid-column: span 3 / span 3;
    }
  }
`;
