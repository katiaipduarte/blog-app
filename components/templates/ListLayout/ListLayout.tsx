import React, { memo, useContext, useEffect, useState } from 'react';
import Link from 'next/link';

import { ListContainer } from './ListLayout.style';
import { BlogConfiguration } from '@constants/blogConfiguration';
import { Post } from '@interfaces/post';
import SearchBar from '@components/ui/SearchBar/SearchBar';
import { PostsContext } from '@lib/context/posts';

type Props = {
  homepage: boolean;
};

const ListLayout = (props: Props): JSX.Element => {
  const { homepage } = props;
  const [postList, setPostList] = useState<Post[]>([]);
  const [term, setTerm] = useState<string>('');
  const { loading, posts } = useContext(PostsContext);

  useEffect(() => {
    if (!loading) {
      setPostList(posts);
    }
  }, [loading, posts]);

  const renderContent = (): JSX.Element => {
    if (postList.length === 0) {
      return (
        <li>
          <h2>No posts found.</h2>
        </li>
      );
    } else {
      return buildPostList();
    }
  };

  const buildPostList = (): JSX.Element => {
    const config: number = homepage
      ? +BlogConfiguration.MAX_HOMEPAGE
      : +BlogConfiguration.MAX_PER_PAGE;

    return (
      <>
        {postList
          .filter((post: Post) => {
            return post.title.toLowerCase().includes(term.toLowerCase());
          })
          .slice(0, config)
          .map((post: Post) => {
            const date = new Date(post.publish_date).toLocaleDateString('en', {
              year: 'numeric',
              month: 'long',
              day: 'numeric'
            });

            return (
              <li key={post.slug} className={homepage ? '' : 'blog-page'}>
                <article>
                  <dl>
                    <dd>
                      <time
                        aria-label={`Published on ${date}`}
                        dateTime={post.publish_date}
                        tabIndex={0}
                      >
                        {date}
                      </time>
                    </dd>
                  </dl>
                  <div className="post">
                    <div className="description">
                      <h2 role="button">
                        <Link
                          href={`/blog/${post.id}/${post.slug}`}
                          aria-label={`Read "${post.title}"`}
                        >
                          {post.title}
                        </Link>
                      </h2>
                      <p role="text" tabIndex={0}>
                        {post.description}
                      </p>
                    </div>
                    {homepage && (
                      <Link
                        href={`/blog/${post.id}/${post.slug}`}
                        aria-label={`Read more about "${post.title}"`}
                      >
                        <button className="transparent" role="button">
                          Read more &rarr;
                        </button>
                      </Link>
                    )}
                  </div>
                </article>
              </li>
            );
          })}
      </>
    );
  };

  const renderAllPostBtn = (): JSX.Element => {
    return (
      <button className="all-post-btn transparent" role="button">
        <Link href="/blog">All Posts &rarr;</Link>
      </button>
    );
  };

  return (
    <ListContainer>
      <div className="title">
        <h1 tabIndex={0}>{homepage ? 'Latest' : 'All Posts'}</h1>
        {homepage && (
          <p role="text" tabIndex={0}>
            Welcome to our Blog App!
          </p>
        )}
        {!homepage && <SearchBar find={(term: string) => setTerm(term)} />}
      </div>
      <ul>{renderContent()}</ul>
      {homepage && renderAllPostBtn()}
    </ListContainer>
  );
};

export default memo(ListLayout);
