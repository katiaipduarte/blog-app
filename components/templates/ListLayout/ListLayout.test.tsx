import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import ListLayout from './ListLayout';

describe('the <ListLayout /> component', () => {
  it('should render post list layout for homepage', () => {
    const tree = renderer.create(<ListLayout homepage={true} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render post list layout for blog page', () => {
    const tree = renderer.create(<ListLayout homepage={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
