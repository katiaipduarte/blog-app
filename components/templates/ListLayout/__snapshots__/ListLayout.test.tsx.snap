// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the <ListLayout /> component should render post list layout for blog page 1`] = `
.c1 .title {
  padding: 1.5rem 0;
  border-bottom: 2px solid;
}

.c1 .title h1 {
  line-height: 2.25rem;
  font-weight: 800;
  font-size: 1.8rem;
}

.c1 .title p {
  font-size: 1.25rem;
  line-height: 1.75rem;
}

.c1 ul li {
  padding: 1.5rem 0 3rem;
}

.c0 ul li:not(.c0 ul li:last-child):not(.c0 ul li.blog-page) {
  border-bottom: 2px solid;
}

.c1 ul li article dl dd {
  line-height: 1.5rem;
  font-size: 1rem;
}

.c1 ul li article .post .description h2 {
  font-wight: 700;
  line-height: 2rem;
  font-size: 1.5rem;
  -webkit-letter-spacing: 0.025rem;
  -moz-letter-spacing: 0.025rem;
  -ms-letter-spacing: 0.025rem;
  letter-spacing: 0.025rem;
}

.c1 ul li article .post .description p {
  font-size: 1rem;
  line-height: 1.75;
  margin-top: 1.5rem;
}

.c1 ul li article .post button {
  margin-top: 1rem;
}

.c1 .all-post-btn {
  float: right;
}

.c2 {
  max-width: 32rem;
  padding: 0.5rem 1rem;
  border-radius: 0.4rem;
  border: 1px solid;
  margin: 1rem 0;
}

.c2:focus-visible,
.c2:focus:not(:focus-visible) {
  outline: auto;
}

.c2 input {
  border: none;
  outline: none;
  font-size: 1rem;
  width: calc(100% - 3rem);
  margin-left: 0.5rem;
  background: transparent;
}

.c2 #search-term-btn,
.c2 #clear-search-term-btn {
  background: none;
  border: none;
  cursor: pointer;
  vertical-align: middle;
}

.c2 #search-term-btn svg,
.c2 #clear-search-term-btn svg {
  height: 1rem;
}

.c2 #search-term-btn:focus,
.c2 #clear-search-term-btn:focus {
  outline: none;
}

@media all and (min-width:700px) {
  .c1 .title {
    padding: 2rem 0;
  }

  .c1 .title h1 {
    line-height: 3.5rem;
    font-size: 3.75rem;
  }

  .c1 .title p {
    margin-top: 1.5rem;
  }

  .c1 article {
    display: grid;
    grid-template-columns: repeat(4,minmax(0,1fr));
    -webkit-align-items: baseline;
    -webkit-box-align: baseline;
    -ms-flex-align: baseline;
    align-items: baseline;
  }

  .c1 .post {
    grid-column: span 3 / span 3;
  }
}

@media (min-width:1080px) {

}

<section
  className="c0 c1"
>
  <div
    className="title"
  >
    <h1
      tabIndex={0}
    >
      All Posts
    </h1>
    <form
      aria-label="Search article by name form"
      className="c2"
      role="searchbox"
      tabIndex={-1}
    >
      <button
        aria-label="Click to search"
        id="search-term-btn"
        onClick={[Function]}
        tabIndex={-1}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <input
        aria-label="Search article by name"
        maxLength={50}
        onChange={[Function]}
        placeholder="Search article by name"
        role="search"
        type="text"
        value=""
      />
      <button
        aria-label="Click to clear search"
        id="clear-search-term-btn"
        onClick={[Function]}
        style={
          Object {
            "visibility": "hidden",
          }
        }
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 "
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </form>
  </div>
  <ul>
    <li>
      <h2>
        No posts found.
      </h2>
    </li>
  </ul>
</section>
`;

exports[`the <ListLayout /> component should render post list layout for homepage 1`] = `
.c1 .title {
  padding: 1.5rem 0;
  border-bottom: 2px solid;
}

.c1 .title h1 {
  line-height: 2.25rem;
  font-weight: 800;
  font-size: 1.8rem;
}

.c1 .title p {
  font-size: 1.25rem;
  line-height: 1.75rem;
}

.c1 ul li {
  padding: 1.5rem 0 3rem;
}

.c0 ul li:not(.c0 ul li:last-child):not(.c0 ul li.blog-page) {
  border-bottom: 2px solid;
}

.c1 ul li article dl dd {
  line-height: 1.5rem;
  font-size: 1rem;
}

.c1 ul li article .post .description h2 {
  font-wight: 700;
  line-height: 2rem;
  font-size: 1.5rem;
  -webkit-letter-spacing: 0.025rem;
  -moz-letter-spacing: 0.025rem;
  -ms-letter-spacing: 0.025rem;
  letter-spacing: 0.025rem;
}

.c1 ul li article .post .description p {
  font-size: 1rem;
  line-height: 1.75;
  margin-top: 1.5rem;
}

.c1 ul li article .post button {
  margin-top: 1rem;
}

.c1 .all-post-btn {
  float: right;
}

@media all and (min-width:700px) {
  .c1 .title {
    padding: 2rem 0;
  }

  .c1 .title h1 {
    line-height: 3.5rem;
    font-size: 3.75rem;
  }

  .c1 .title p {
    margin-top: 1.5rem;
  }

  .c1 article {
    display: grid;
    grid-template-columns: repeat(4,minmax(0,1fr));
    -webkit-align-items: baseline;
    -webkit-box-align: baseline;
    -ms-flex-align: baseline;
    align-items: baseline;
  }

  .c1 .post {
    grid-column: span 3 / span 3;
  }
}

<section
  className="c0 c1"
>
  <div
    className="title"
  >
    <h1
      tabIndex={0}
    >
      Latest
    </h1>
    <p
      tabIndex={0}
    >
      Welcome to our Blog App!
    </p>
  </div>
  <ul>
    <li>
      <h2>
        No posts found.
      </h2>
    </li>
  </ul>
  <button
    className="all-post-btn transparent"
    role="button"
  >
    <a
      href="/blog"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      All Posts â†’
    </a>
  </button>
</section>
`;
